crates/router/src/connector/shift4/transformers.rs

```diff
@@ -29,42 +75,121 @@ pub struct Card {
-                    description: item.description.clone(),
-                    captured: submit_for_settlement,
-                };
-                Ok(payment_request)
+        match &item.request.payment_method_data {
+            api::PaymentMethodData::Card(ccard) => get_card_payment_request(item, ccard),
+            api::PaymentMethodData::BankRedirect(redirect_data) => {
+                get_bank_redirect_request(item, redirect_data)
             }
-            _ => Err(
-                errors::ConnectorError::NotImplemented("Current Payment Method".to_string()).into(),
-            ),
+            _ => Err(errors::ConnectorError::NotImplemented("Payment Method".to_string()).into()),
         }
     }
 }
 
+fn get_card_payment_request(
+    item: &types::PaymentsAuthorizeRouterData,
+    card: &api_models::payments::Card,
+) -> Result<Shift4PaymentsRequest, Error> {
+    let submit_for_settlement = submit_for_settlement(item);
+    let card = Some(Card {
+        number: card.card_number.clone(),
+        exp_month: card.card_exp_month.clone(),
+        exp_year: card.card_exp_year.clone(),
+        cardholder_name: card.card_holder_name.clone(),
+    });
+    Ok(Shift4PaymentsRequest {
+        amount: item.request.amount.to_string(),
+        card,
+        currency: item.request.currency.to_string(),
+        description: item.description.clone(),
+        captured: submit_for_settlement,
+        payment_method: None,
+        flow: None,
+    })
+}
+
+fn get_bank_redirect_request(
+    item: &types::PaymentsAuthorizeRouterData,
+    redirect_data: &payments::BankRedirectData,
+) -> Result<Shift4PaymentsRequest, Error> {
+    let submit_for_settlement = submit_for_settlement(item);
+    let method_type = PaymentMethodType::from(redirect_data);
+    let billing = get_billing(item)?;
+    let payment_method = Some(PaymentMethod {
+        method_type,
+        billing,
+    });
+    let flow = get_flow(item);
+    Ok(Shift4PaymentsRequest {
+        amount: item.request.amount.to_string(),
+        card: None,
+        currency: item.request.currency.to_string(),
+        description: item.description.clone(),
+        captured: submit_for_settlement,
+        payment_method,
+        flow: Some(flow),
+    })
+}
+
+impl From<&payments::BankRedirectData> for PaymentMethodType {
+    fn from(value: &payments::BankRedirectData) -> Self {
+        match value {
+            payments::BankRedirectData::Eps { .. } => Self::Eps,
+            payments::BankRedirectData::Giropay { .. } => Self::Giropay,
+            payments::BankRedirectData::Ideal { .. } => Self::Ideal,
+            payments::BankRedirectData::Sofort { .. } => Self::Sofort,
+        }
+    }
+}
+
+fn get_flow(item: &types::PaymentsAuthorizeRouterData) -> Flow {
+    Flow {
+        return_url: item.router_return_url.clone(),
+    }
+}
+
+fn get_billing(item: &types::PaymentsAuthorizeRouterData) -> Result<Option<Billing>, Error> {
```

Comment:
Mandatory fields required for the payment will be passed in the `payment_method_data`. Should that data not be taken from there?

Reply:
here I am constructing only the `Billing `, from `item.address.billing`. So `payment_method_data` is not required with in this method.